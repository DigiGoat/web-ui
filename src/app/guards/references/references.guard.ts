import { inject } from '@angular/core';
import { CanActivateFn } from '@angular/router';
import { ConfigService } from '../../services/config/config.service';

export const referencesGuard: CanActivateFn = () => {
  return inject(ConfigService).references;
};
