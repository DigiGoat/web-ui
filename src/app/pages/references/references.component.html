<app-goats [goats]="getter" [name]="name"></app-goats>