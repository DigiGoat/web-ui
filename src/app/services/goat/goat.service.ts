import { Observable } from 'rxjs';
import { retry } from 'rxjs/operators';

import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';


@Injectable({
  providedIn: 'root'
})
export class GoatService {
  constructor(private http: HttpClient) { }
  private _does: Goat[] = [];
  public does = new Observable<Goat[]>(observer => {
    if (this._does.length) {
      console.debug('Loaded Does From Cache', this._does);
      observer.next(this._does);
    } else {
      this.http.get<Goat[]>('/assets/resources/does.json')
        .pipe(
          retry(3), // retry a failed request up to 3 times
        )
        .subscribe({
          next: data => {
            this._does = data;
            console.debug('Loaded Does From Server', data);
            observer.next(data);
          },
          error: err => {
            if (err.status === 0) {
              // A client-side or network error occurred. Handle it accordingly.
              console.warn('An error occurred:', err.error);
            } else {
              // The backend returned an unsuccessful response code.
              // The response body may contain clues as to what went wrong.
              console.warn(
                `Backend returned code ${err.status}, body was: `, err.error);
            }
            // Return an observable with a user-facing error message.
            observer.error(err);
          }
        });
    }
  });
  private _bucks: Goat[] = [];
  public bucks = new Observable<Goat[]>(observer => {
    if (this._bucks.length) {
      console.debug('Loaded Bucks From Cache', this._bucks);
      observer.next(this._bucks);
    } else {
      this.http.get<Goat[]>('/assets/resources/bucks.json')
        .pipe(
          retry(3), // retry a failed request up to 3 times
        )
        .subscribe({
          next: data => {
            this._bucks = data;
            console.debug('Loaded Bucks From Server', data);
            observer.next(data);
          },
          error: err => {
            if (err.status === 0) {
              // A client-side or network error occurred. Handle it accordingly.
              console.warn('An error occurred:', err.error);
            } else {
              // The backend returned an unsuccessful response code.
              // The response body may contain clues as to what went wrong.
              console.warn(
                `Backend returned code ${err.status}, body was: `, err.error);
            }
            // Return an observable with a user-facing error message.
            observer.error(err);
          }
        });
    }
  });
}
//export type Goat = (OwnedGoats['result']['items'][number] & { nickname: string; description: string; awards: Awards['result']['items']; colorAndMarking: string; /*obtained?: string;*/ });
export type Goat = {
  nickname?: string;
  name: string;
  description?: string;
  normalizeId: string;
  dateOfBirth: string;
  colorAndMarking?: string;
  animalTattoo?: [{ tattoo?: string; tattooLocation?: { name?: string; }; }, { tattoo?: string; tattooLocation?: { name?: string; }; }];
};
export const Goat = {
  nickname: 'Your Goats Farm Name',
  name: 'YOUR GOATS ADGA NAME',
  description: 'A quick description about your goat',
  normalizeId: 'PD12345',
};
