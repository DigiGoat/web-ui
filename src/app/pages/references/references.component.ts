import { Component } from '@angular/core';
import { GoatService } from '../../services/goat/goat.service';

@Component({
    selector: 'app-references',
    templateUrl: './references.component.html',
    styleUrl: './references.component.scss',
    standalone: false
})
export class ReferencesComponent {
  name = 'References';
  getter = this.goatService.references;
  constructor(public goatService: GoatService) { }
}
