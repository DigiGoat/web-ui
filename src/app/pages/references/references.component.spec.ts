import { Observable } from 'rxjs';
import { of } from 'rxjs/internal/observable/of';

import { ComponentFixture, TestBed } from '@angular/core/testing';

import { GoatService } from '../../services/goat/goat.service';
import { ReferencesComponent } from './references.component';


jest.mock('../../services/goat/goat.service');
const goatService = jest.mocked(GoatService);
describe('ReferencesComponent', () => {
  let component: ReferencesComponent;
  let fixture: ComponentFixture<ReferencesComponent>;

  beforeEach(async () => {
    goatService.mockClear();
    await TestBed.configureTestingModule({
      declarations: [ReferencesComponent],
      providers: [GoatService],
    })
      .compileComponents();
    fixture = TestBed.createComponent(ReferencesComponent);
    component = fixture.componentInstance;
    TestBed.inject(GoatService);

    goatService.mockImplementation(() => {
      return {
        references: of([])
      } as unknown as GoatService;

    });
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should call GoatService once', () => {
    expect(goatService).toHaveBeenCalledTimes(1);
  });
  it('should specify a name of References', () => {
    expect(component.name).toBe('References');
  });
  it('should specify an Observable getter of references', () => {
    expect(component.getter).toBeInstanceOf(Observable);
  });
  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
