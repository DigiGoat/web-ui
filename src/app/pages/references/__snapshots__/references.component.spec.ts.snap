// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferencesComponent should match the snapshot 1`] = `
<app-references
  getter={[Function Observable]}
  goatService={[Function Object]}
  name={[Function String]}
>
  <app-goats />
</app-references>
`;
